src/query/tsdb/remote/server.go

```diff
@@ -185,22 +210,31 @@ func (s *grpcServer) CompleteTags(
 	fetchOpts := storage.NewFetchOptions()
 	fetchOpts.Limit = s.queryContextOpts.LimitMaxTimeseries
 
-	results, err := s.storage.CompleteTags(ctx, completeTagsQuery, fetchOpts)
+	completed, err := s.storage.CompleteTags(ctx, completeTagsQuery, fetchOpts)
 	if err != nil {
 		logger.Error("unable to complete tags", zap.Error(err))
 		return err
 	}
 
-	response, err := encodeToCompressedCompleteTagsResult(results)
-	if err != nil {
-		logger.Error("unable to encode complete tags result", zap.Error(err))
-		return err
-	}
-
-	err = stream.SendMsg(response)
-	if err != nil {
-		logger.Error("unable to send complete tags result", zap.Error(err))
+	tags := completed.CompletedTags
+	size := min(defaultBatch, len(tags))
+	for ; len(tags) > size; tags = tags[size:] {
+		results := &storage.CompleteTagsResult{
+			CompleteNameOnly: completed.CompleteNameOnly,
+			CompletedTags:    tags,
+		}
+
+		response, err := encodeToCompressedCompleteTagsResult(results)
+		if err != nil {
+			logger.Error("unable to encode complete tags result", zap.Error(err))
+			return err
+		}
+
+		err = stream.SendMsg(response)
+		if err != nil {
+			logger.Error("unable to send complete tags result", zap.Error(err))
```

Comment:
Here too, should `return err` here if send fails, it's quite likely the connection is broken and should just terminate the send.